core/google/api/core/gapic_v1/method.py

```diff
@@ -0,0 +1,222 @@
+from google.api.core import timeout
+from google.api.core.helpers import grpc_helpers
+
+_PY_VERSION = platform.python_version()
+_GRPC_VERSION = pkg_resources.get_distribution('grpcio').version
+_API_CORE_VERSION = pkg_resources.get_distribution('google-cloud-core').version
+METRICS_METADATA_KEY = 'x-goog-api-client'
+
+
+def _is_not_none_or_false(value):
+    return value is not None and value is not False
+
+
+def _apply_decorators(func, decorators):
+    """Apply a list of decorators to a given function.
+
+    ``decorators`` may contain items that are ``None`` or ``False`` which will
+    be ignored.
+    """
+    decorators = list(filter(_is_not_none_or_false, decorators))
+
+    for decorator in reversed(decorators):
+        func = decorator(func)
+
+    return func
+
+
+def _prepare_metadata(metadata):
+    """Transforms metadata to gRPC format and adds global metrics.
+
+    Args:
+        metadata (Optional[Mapping[str, str]]): Any current metadata.
+
+    Returns:
+        Sequence[Tuple(str, str)]: The gRPC-friendly metadata keys and values.
+    """
+    if metadata is None:
+        metadata = {}
+
+    client_metadata = 'api-core/{} gl-python/{} grpc/{}'.format(
+        _API_CORE_VERSION, _PY_VERSION, _API_CORE_VERSION)
+
+    # Merge this with any existing metric metadata.
+    if METRICS_METADATA_KEY in metadata:
+        client_metadata = '{} {}'.format(
+            client_metadata, metadata[METRICS_METADATA_KEY])
+
+    metadata[METRICS_METADATA_KEY] = client_metadata
+
+    return list(metadata.items())
+
+
+class _GapicCallable(object):
+    """Callable that applies retry, timeout, and metadata logic."""
+
+    def __init__(self, target, retry, timeout, metadata):
+        """
+        Args:
+            target (Callable): The low-level RPC method. If timeout is not
+                False, should accept a timeout argument. If metadata is not
+                False, it should accept a metadata argument.
+            retry (google.api.core.retry.Retry): The default retry for the
+                callable. If False, this callable will not retry.
+            timeout (google.api.core.timeout.Timeout): The default timeout
+                for the callable. If False, this callable will not specify
+                a timeout argument to the low-level RPC method.
+            metadata (Sequence[Tuple[str, str]]): gRPC call metadata that's
+                passed to the low-level RPC method. If False, this callable
+                will not specify a metadata argument to the method.
+        """
+        self._target = target
+        self._retry = retry
+        self._timeout = timeout
+        self._metadata = metadata
+
+    def __call__(self, *args, **kwargs):
+        """Invoke the low-level RPC with retry, timeout, and metadata."""
+        # Note: Due to Python 2 lacking keyword-only arguments we use this
+        # pattern to extract the retry and timeout params.
+        retry_ = kwargs.pop('retry', None)
```

Comment:
Trivial question irrelevant to code review: Why the trailing underscore on `retry_` (and `timeout_`)?

Reply:
Because I import `retry` and `timeout`.